src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs

```diff
@@ -28,4 +32,57 @@ internal interface ITextBufferVisibilityChangedCallback
     {
         void OnTextBufferVisibilityChanged();
     }
+
+    internal static class ITextBufferVisibilityTrackerExtensions
+    {
+        /// <summary>
+        /// Waits the specified amount of time while the specified <paramref name="subjectBuffer"/> is not visible.  If any
+        /// document visibility changes happen, the delay will cancel.
+        /// </summary>
+        public static async Task DelayWhileNonVisibleAsync(
+            this ITextBufferVisibilityTracker? service,
+            IThreadingContext threadingContext,
+            ITextBuffer subjectBuffer,
+            TimeSpan timeSpan,
+            CancellationToken cancellationToken)
+        {
+            // Only add a delay if we have access to a service that will tell us when the buffer become visible or not.
+            if (service is null)
+                return;
+
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            if (service.IsVisible(subjectBuffer))
+                return;
+
+            // ensure we listen for visibility changes before checking.  That way we don't have a race where we check
+            // something see it is not visible, but then do not hear about its visibility change because we've hooked up
+            // our event after that happens.
+            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
+            var callback = new VisibilityCallback(visibilityChangedTaskSource);
```

Comment:
so this brings me to a question on the original API - any reason why the callback isn't a func or something similar?  Seems like each consumer needs a bit more boilerplate to use the ITextBufferVisibilityChangedCallback